packages/utils/src/sha256.cairo

```diff
@@ -34,10 +35,340 @@ pub fn double_sha256_u32_array(words: Array<u32>) -> Digest {
     DigestTrait::new(compute_sha256_u32_array(input2, 0, 0))
 }
 
+/// Cairo implementation of the corelib `compute_sha256_byte_array` function.
+fn compute_sha256_byte_array(arr: @ByteArray) -> [u32; 8] {
+    let mut sha_input: Array<u8> = array![];
+
+    let mut i: usize = 0;
+    while i != arr.len() {
+        sha_input.append(arr[i]);
+        i += 1;
+    };
+
+    let hash = sha256(sha_input);
+
+    u8_array_to_u32_8_fixed_size_array(hash)
+}
+
+/// Cairo implementation of the corelib `compute_sha256_u32_array` function.
+fn compute_sha256_u32_array(
+    mut input: Array<u32>, last_input_word: u32, last_input_num_bytes: u32
+    ) -> [
+    u32
+; 8] {
+    let mut sha_input: Array<u8> = array![];
+
+    for value in input
+        .span() {
+            sha_input.append((shr(*value, 24_u32) & 0xFF_u32).try_into().unwrap());
+            sha_input.append((shr(*value, 16_u32) & 0xFF_u32).try_into().unwrap());
+            sha_input.append((shr(*value, 8_u32) & 0xFF_u32).try_into().unwrap());
+            sha_input.append((*value & 0xFF_u32).try_into().unwrap());
+        };
+
+    if last_input_num_bytes > 0 {
+        for i in 0
+            ..last_input_num_bytes {
+                let shift = 24_u32 - i * 8_u32;
+                sha_input.append((shr(last_input_word, shift) & 0xFF_u32).try_into().unwrap());
+            }
+    }
+
+    let hash = sha256(sha_input);
+
+    u8_array_to_u32_8_fixed_size_array(hash)
+}
+
+/// Converts an array of u8 to a fixed size array of 8 u32.
+fn u8_array_to_u32_8_fixed_size_array(input: Array<u8>) -> [u32; 8] {
+    let mut arr: Array<u32> = array![];
+    let mut i = 0;
+    while i != input.len() {
+        let a: u32 = (*input[i]).into();
+        let b: u32 = (*input[i + 1]).into();
+        let c: u32 = (*input[i + 2]).into();
+        let d: u32 = (*input[i + 3]).into();
+
+        let value = shl(a, 24_u32) | shl(b, 16_u32) | shl(c, 8_u32) | d;
+
+        arr.append(value);
+        i += 4;
+    };
+
+    [*arr[0], *arr[1], *arr[2], *arr[3], *arr[4], *arr[5], *arr[6], *arr[7]]
+}
+
+/// Cairo implementation of sha256 hash function.
+fn sha256(mut data: Array<u8>) -> Array<u8> {
+    let data_len: u64 = (data.len() * 8).into();
+
+    // add one
```

Comment:
I wonder if we can adapt `add_sha256_padding` from: https://github.com/starkware-libs/cairo/blob/3eb7af7222c42ba56a886cbbd7cd5615d6ad36d6/corelib/src/sha256.cairo#L72 and get rid of u8 -> u32 conversions altogether.

Reply:
Ahah, we are discussing the same thing in TG with @TAdev0 😄